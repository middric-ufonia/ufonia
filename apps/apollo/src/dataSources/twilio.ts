import Twilio from "twilio/lib/rest/Twilio";

export class TwilioDataSource {
  private client;
  private telNo;

  constructor(
    client: Twilio,
    options: {
      telNo: string;
    }
  ) {
    this.client = client;
    this.telNo = options.telNo;
  }
  async createCall(message: string, recipient: string) {
    return await this.client.calls.create({
      twiml: `<Response><Say>${message}</Say></Response>`,
      to: recipient,
      from: this.telNo,
    });
  }
}
