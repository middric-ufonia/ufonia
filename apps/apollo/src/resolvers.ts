import { GraphQLError } from "graphql";
import type { Resolvers } from "./@types/graphQLSchema.js";
import { UfoniaApolloServerContext } from "./index.js";

export const resolvers: Resolvers = {
  Query: {
    healthCheck: () => true,
  },
  Mutation: {
    sendMessage: async (
      _parent,
      args,
      { dataSources }: UfoniaApolloServerContext
    ) => {
      try {
        const call = await dataSources.twilio.createCall(
          args.input.msg,
          args.input.tel
        );

        return { sid: call.sid, code: "200", success: true };
      } catch (error) {
        throw new GraphQLError("Unable to connect to service provider", {
          extensions: {
            code: "INTERNAL_SERVER_ERROR",
          },
          originalError: error,
        });
      }
    },
  },
};
