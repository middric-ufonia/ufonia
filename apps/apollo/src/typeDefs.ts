export const typeDefs = `#graphql
  type Message {
    tel: String!
    msg: String!
  }

  type SendMessageResponse {
    code: String!
    success: Boolean!
    sid: String
  }

  input SendMessageInput {
    tel: String!
    msg: String!
  }

  type Query {
    healthCheck: Boolean!
  }

  type Mutation {
    sendMessage(input: SendMessageInput): SendMessageResponse
  }
`;
