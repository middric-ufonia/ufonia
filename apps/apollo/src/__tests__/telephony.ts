import { ApolloServer } from "@apollo/server";
import assert from "assert";
import { UfoniaApolloServerContext } from "..";
import { TwilioDataSource } from "../dataSources/twilio";
import { resolvers } from "../resolvers";
import { typeDefs } from "../typeDefs";

describe("Telephony", () => {
  let testServer: ApolloServer<UfoniaApolloServerContext>;

  beforeAll(() => {
    testServer = new ApolloServer({
      typeDefs,
      resolvers,
    });
  });

  it("returns a successful healthcheck", async () => {
    const response = await testServer.executeOperation({
      query: "query HealthCheck { healthCheck }",
    });
    assert(response.body.kind === "single");
    expect(response.body.singleResult.errors).toBeUndefined();
    expect(response.body.singleResult.data?.healthCheck).toBeTruthy();
  });

  it("makes a phone call using the Twilio datasource", async () => {
    const twilio = new TwilioDataSource(null, {
      telNo: "",
    });
    twilio.createCall = jest.fn(async (message: string, recipient: string) => {
      return await { sid: "sid" };
    });

    const response = await testServer.executeOperation(
      {
        query: `#graphql
          mutation SendMessage($input: SendMessageInput) {
            sendMessage(input: $input) {
              sid
              code
              success
            }
          }`,
        variables: {
          input: {
            tel: "123",
            msg: "abc",
          },
        },
      },
      {
        contextValue: {
          dataSources: {
            twilio,
          },
        },
      }
    );
    assert(response.body.kind === "single");
    expect(response.body.singleResult.errors).toBeUndefined();
    expect(response.body).toMatchSnapshot();
  });
});
