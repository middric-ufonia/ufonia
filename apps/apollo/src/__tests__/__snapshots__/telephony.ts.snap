// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telephony makes a phone call using the Twilio datasource 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "sendMessage": {
        "code": "200",
        "sid": "sid",
        "success": true,
      },
    },
    "errors": undefined,
  },
}
`;
